### ISSUE TITLE
[Feature Request

### ISSUE THREAD

**Issue by jeqo** [openzipkin/brave#895]
What kind of issue is this?

 - [ ] Question. This issue tracker is not the place for questions. If you want to ask how to do
       something, or to understand why something isn't working the way you expect it to, use Gitter
       or Stack Overflow. https://gitter.im/openzipkin/zipkin https://stackoverflow.com/questions/tagged/zipkin

 - [ ] Bug report. If you’ve found a bug, spend the time to write a failing test. Bugs with tests
       get fixed and stay fixed. If you have a solution in mind, skip raising an issue and open a
       pull request instead.

 - [ ] Feature Request. First, look at existing issues to see if the feature has been requested
       before. If you don't find anything, tell us what problem you’re trying to solve. Often a
       solution already exists! Don’t send pull requests to implement new features without first
       getting our support. Sometimes we leave features out on purpose to keep the project small.

**Comment by adriancole** [comment]
I think our default naming is similar for mysql, though we don't include any sort of append logic based on user input, rather if they don't put anything, the default is a pattern like this.

I'm tempted to make this a pattern, but I can see the value in just concatenating as then you don't need to worry about variable naming of queue vs topic etc.

**Comment by shakuzen** [comment]
The analog of topic/queue to RPC tracing seems to be method/endpoint, which we generally put in the span name, not the service. I know it isn't an apples-to-apples comparison, but I'm wondering why we wouldn't prefer to do that with messaging as well.

Remote service name to me seems better suited to identifying a kafka/rabbitmq cluster than a topic/exchange. The dependency graph was mentioned in the description, and for using this in the dependency graph, topic/queue seems quite high cardinality (like HTTP endpoints would likely be) compared to a cluster identifier. I would also think people would want to identify broker clusters because if there is a failure/outage in the cluster it will affect all topics/exchanges, in general.

**Comment by jeqo** [comment]
@shakuzen the main difference between messaging to RPC is that we don't have the server (broker) span to name it. To get there, we would need to instrument the broker side. 

Dependency graph is the main driver to me: is true that we will get more services (x #topics) but instead of getting all producers/consumers pointing to a cluster, we will get producer/consumer per topic pointing to cluster+topicName instead which seems more useful in terms of understanding but also for metrics.

I agree that identifying cluster is important as well, but not sure remoteServiceName alone capture failure/outage neither as value is static. I was looking into how to add host:port to remote endpoint via Kafka AdminClient, and seems possible with `AdminClient#describeTopics` (https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/admin/AdminClient.html) I will check this further and create another issue if it make sense.

@adriancole I like the pattern idea, but it could be a bit harder to configure/use as you said. I was thinking to include a separator variable between service name and topic name for instance to make it easier to distinguish.

**Comment by adriancole** [comment]
jorge, would a finishedspanhandler work? you should have the topic names
there right?

**Comment by shakuzen** [comment]
> Dependency graph is the main driver to me: is true that we will get more services (x #topics) but instead of getting all producers/consumers pointing to a cluster, we will get producer/consumer per topic pointing to cluster+topicName instead which seems more useful in terms of understanding but also for metrics.

I agree with this statement, but I think the same argument applies for HTTP endpoints, which people have asked to see in the dependency graph in the past. So I'm just wondering if we shouldn't think about this and its implications in a more generic way before we put something out in the wild. Maybe not though. I don't have a concrete proposal.

**Comment by shakuzen** [comment]
Also a note about RabbitMQ: producers publish to an exchange, and then depending on RabbitMQ's configuration, the message ends up in zero, one, or multiple queues. That is to say the producer only knows the exchange it is publishing to, and the consumer only knows the queue it is consuming from. So I'm not sure how this would be reconciled on the dependency graph.

**Comment by jeqo** [comment]
@adriancole yes, topic names will be available via tag I think. If it's not part of the instrumentation, finishedspanhandler could work, or setting remoteservicename manually as well.

> So I'm just wondering if we shouldn't think about this and its implications in a more generic way before we put something out in the wild. Maybe not though. I don't have a concrete proposal.

@shakuzen this might be true, I've been wondering about how to do path analysis by having a service:span call graph instead of only services. Similar to path aggregation described here: https://cwiki.apache.org/confluence/display/ZIPKIN/2018-07-18+Aggregation+and+Analysis+at+Netflix

Even then, the lack of span on the remote service (when messaging) will require to use tags from the producer/consumer span to know which topic was used. Thinking out loud: Should be possible to define remote service span names from prod/consumers when tracing messaging apps?

**Comment by adriancole** [comment]
I think somewhere we had an issue to create something for messaging like
we have for http (HttpParser etc). There are probably not many low
cardinality inputs available for use in naming etc, however, having access
to the message object to make a decision should be helpful. I think it is
time to implement that abstraction, but different from Http ... add the
remoteServiceName aspect. In fact we probably should also add
remoteServiceName to the http parsers :P

Are you up to starting this? I think it is long overdue.

**Comment by jeqo** [comment]
@adriancole I remember we discussed something similar around jms instrumentation. 
Do you mean to have a `brave.messaging` layer to reuse on jms/kafka/..., right? I like the challenge, I will play around to see how it will look like based on `brave.http` module, this might make it easier to implement this idea over messaging modules.

About remoteServiceName for http, I will take a look as well and create an issue to track it.

**Comment by adriancole** [comment]
yes exactly this
